aten/src/ATen/native/Loss.cpp

```diff
@@ -76,27 +76,43 @@ Tensor margin_ranking_loss(const Tensor& input1, const Tensor& input2, const Ten
   return apply_loss_reduction(output, reduction);
 }
 
-Tensor kl_div(const Tensor& input, const Tensor& target, int64_t reduction) {
+Tensor _kl_div_log_target(const Tensor& input, const Tensor& target, int64_t reduction) {
+  auto output = at::exp(target) * (target - input);
+  return apply_loss_reduction(output, reduction);
+}
+
+Tensor _kl_div_non_log_target(const Tensor& input, const Tensor& target, int64_t reduction) {
   auto output_pos = target * (at::log(target) - input);
   auto zeros = at::zeros_like(output_pos, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
   auto output = at::where(target > 0, output_pos, zeros);
   return apply_loss_reduction(output, reduction);
 }
 
-Tensor kl_div_backward_cpu(const Tensor& grad, const Tensor& input, const Tensor& target, int64_t reduction) {
+Tensor kl_div(const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
+  return log_target ? _kl_div_log_target(input, target, reduction)
+                    : _kl_div_non_log_target(input, target, reduction);
+}
+
+Tensor kl_div_backward_cpu(const Tensor& grad, const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
   auto grad_input = at::zeros_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
   auto grad_expand = grad.expand_as(input);
-  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "kl_div_backward_cpu", [&]() {
-    at::CPU_tensor_apply3<scalar_t, scalar_t, scalar_t>(
-        grad_input,
-        target,
-        grad_expand,
-        [] (scalar_t& grad_input_val, const scalar_t& target_val, const scalar_t& grad_val) {
-          if (target_val > 0) {
-            grad_input_val = -target_val * grad_val;
-          }
-        });
-  });
+  if (!log_target) {
+    AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "kl_div_backward_cpu", [&]() {
+      at::CPU_tensor_apply3<scalar_t, scalar_t, scalar_t>(
+          grad_input,
+          target,
+          grad_expand,
+          [] (scalar_t& grad_input_val, const scalar_t& target_val, const scalar_t& grad_val) {
+            if (target_val > 0) {
+              grad_input_val = -target_val * grad_val;
+            }
+          });
+    });
+  }
+  else {
+    grad_input = -at::exp(target) * grad_expand;
```

Comment:
Are you sure you don't want to fuse these operations too?

Reply:
I did not fuse because `backward` is not fused. Shall I make both forward and backward fused?

Reply:
I don't know if the fusion matters. But your hypothesis for why it was fused originally is very plausible.